File: python/add_expense_dialog.py
from __future__ import annotations
from pathlib import Path
from typing import Any
from PySide6.QtCore import QFile, QDate
from PySide6.QtUiTools import QUiLoader
from PySide6.QtWidgets import (
    QDateEdit,
    QDialog,
    QDoubleSpinBox,
    QLineEdit,
    QMessageBox,
    QPushButton,
    QVBoxLayout,
    QWidget,
)



class AddEditDialog(QDialog):
    def __init__(self, parent=None, expense: dict[str, Any] | None = None) -> None:
        super().__init__(parent)
        ui_path = Path(__file__).resolve().parent.parent / "shared" / "ui" / "add_expense_dialog.ui"
        loader = QUiLoader()
        ui_file = QFile(str(ui_path))
        if not ui_file.open(QFile.ReadOnly):
            raise RuntimeError(f"Unable to open UI file: {ui_path}")
        dialog_widget = loader.load(ui_file)
        ui_file.close()
        if dialog_widget is None:
            raise RuntimeError(f"Unable to load UI file: {ui_path}")
        if not isinstance(dialog_widget, QWidget):
            raise RuntimeError(f"Unexpected root widget in UI file: {ui_path}")

        host_layout = QVBoxLayout(self)
        host_layout.setContentsMargins(0, 0, 0, 0)
        host_layout.addWidget(dialog_widget)

        self.dateEdit = dialog_widget.findChild(QDateEdit, "dateEdit")
        self.amountSpinBox = dialog_widget.findChild(QDoubleSpinBox, "amountSpinBox")
        self.categoryLineEdit = dialog_widget.findChild(QLineEdit, "categoryLineEdit")
        self.descriptionLineEdit = dialog_widget.findChild(QLineEdit, "descriptionLineEdit")
        self.saveButton = dialog_widget.findChild(QPushButton, "saveButton")
        self.cancelButton = dialog_widget.findChild(QPushButton, "cancelButton")

        if (
            self.dateEdit is None
            or self.amountSpinBox is None
            or self.categoryLineEdit is None
            or self.descriptionLineEdit is None
            or self.saveButton is None
            or self.cancelButton is None
        ):
            raise RuntimeError("add_expense_dialog.ui is missing one or more required widgets")

        self.setWindowTitle("Add Expense" if expense is None else "Edit Expense")
        self.saveButton.clicked.connect(self.on_save)
        self.cancelButton.clicked.connect(self.reject)

        self.dateEdit.setDate(QDate.currentDate())
        if expense is not None:
            y, m, d = expense["date"].split("-")
            self.dateEdit.setDate(QDate(int(y), int(m), int(d)))
            self.amountSpinBox.setValue(float(expense["amount"]))
            self.categoryLineEdit.setText(expense["category"])
            self.descriptionLineEdit.setText(expense["description"])

    def on_save(self) -> None:
        if not self.categoryLineEdit.text().strip():
            QMessageBox.warning(self, "Validation Error", "category is required")
            return
        if not self.descriptionLineEdit.text().strip():
            QMessageBox.warning(self, "Validation Error", "description is required")
            return
        self.accept()

    def get_values(self) -> dict[str, Any]:
        return {
            "date": self.dateEdit.date().toString("yyyy-MM-dd"),
            "category": self.categoryLineEdit.text().strip(),
            "description": self.descriptionLineEdit.text().strip(),
            "amount": float(self.amountSpinBox.value()),
        }

File: cpp/src/add_expense_dialog.cpp
#include "add_expense_dialog.h"
#include "ui_add_expense_dialog.h"
#include <QMessageBox>

AddExpenseDialog::AddExpenseDialog(QWidget* parent)
    : QDialog(parent)
    , m_ui(new Ui::AddExpenseDialog)   // allocate on the heap with 'new'
    , m_editId(-1)
    , m_editUserId(-1)
{
    m_ui->setupUi(this);
    setWindowTitle("Add Expense");

    m_ui->dateEdit->setDate(QDate::currentDate());
    m_ui->dateEdit->setCalendarPopup(true);
    m_ui->amountSpinBox->setMinimum(0.01);
    m_ui->amountSpinBox->setMaximum(999999.99);
    m_ui->amountSpinBox->setDecimals(2);
    m_ui->amountSpinBox->setPrefix("$ ");

    connect(m_ui->saveButton,   &QPushButton::clicked, this, &AddExpenseDialog::onSave);
    connect(m_ui->cancelButton, &QPushButton::clicked, this, &QDialog::reject);
}

// Destructor — called automatically when the object is destroyed.
// We must delete m_ui here because we created it with 'new'.
// Python never needs this — the garbage collector handles it.
AddExpenseDialog::~AddExpenseDialog()
{
    delete m_ui;
}

void AddExpenseDialog::loadRecord(const ExpenseRecord& r)
{
    m_editId     = r.id;
    m_editUserId = r.userId;
    setWindowTitle("Edit Expense");
    m_ui->dateEdit->setDate(r.date);
    m_ui->amountSpinBox->setValue(r.amount);
    m_ui->categoryLineEdit->setText(r.category);
    m_ui->descriptionLineEdit->setText(r.description);
}

ExpenseRecord AddExpenseDialog::getRecord() const
{
    return ExpenseRecord(
        m_editId, m_editUserId,
        m_ui->dateEdit->date(),
        m_ui->amountSpinBox->value(),
        m_ui->categoryLineEdit->text().trimmed(),
        m_ui->descriptionLineEdit->text().trimmed());
}

void AddExpenseDialog::onSave()
{
    if (validate()) accept();
}

bool AddExpenseDialog::validate()
{
    if (m_ui->categoryLineEdit->text().trimmed().isEmpty())
    {
        QMessageBox::warning(this, "Validation", "Please enter a category.");
        return false;
    }
    if (m_ui->amountSpinBox->value() <= 0.0)
    {
        QMessageBox::warning(this, "Validation", "Amount must be greater than zero.");
        return false;
    }
    return true;
}
